controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/UpdateReaderGroupTask.java

```diff
@@ -68,30 +69,39 @@ public UpdateReaderGroupTask(final StreamMetadataTasks streamMetaTasks,
                 }
                 return streamMetadataStore.getReaderGroupConfigRecord(scope, readerGroup, context, executor)
                        .thenCompose(rgConfigRecord -> {
-                       if (rgConfigRecord.getObject().getGeneration() != generation) {
-                           log.warn("Skipping processing of Reader Group update request {} as generation did not match.", requestId);
-                           return CompletableFuture.completedFuture(null);
-                       }
-                       if (rgConfigRecord.getObject().isUpdating() &&
-                          (!ReaderGroupConfig.StreamDataRetention.values()[rgConfigRecord.getObject().getRetentionTypeOrdinal()]
-                            .equals(ReaderGroupConfig.StreamDataRetention.NONE))) {
-                            // update Stream metadata tables, only if RG is a Subscriber
-                            Iterator<String> streamIter = rgConfigRecord.getObject().getStartingStreamCuts().keySet().iterator();
-                            String scopedRGName = NameUtils.getScopedReaderGroupName(scope, readerGroup);
-                            Iterator<String> removeStreamsIter = streamsToBeUnsubscribed.stream().iterator();
-                            return Futures.loop(() -> streamIter.hasNext(), () -> {
-                                          Stream stream = Stream.of(streamIter.next());
-                                          return streamMetadataStore.addSubscriber(stream.getScope(),
-                                                       stream.getStreamName(), scopedRGName, rgConfigRecord.getObject().getGeneration(),
-                                                  null, executor);
-                                      }, executor)
-                                    .thenCompose(v -> Futures.loop(() -> removeStreamsIter.hasNext(), () -> {
-                                        Stream stream = Stream.of(removeStreamsIter.next());
-                                        return streamMetadataStore.deleteSubscriber(stream.getScope(),
-                                                stream.getStreamName(), scopedRGName, rgConfigRecord.getObject().getGeneration(), null, executor);
-                                    }, executor))
-                                    .thenCompose(v -> streamMetadataStore.completeRGConfigUpdate(scope, readerGroup, rgConfigRecord, context, executor));
-                            }
+                           if (rgConfigRecord.getObject().getGeneration() != generation) {
+                               log.warn("Skipping processing of Reader Group update request {} as generation did not match.", requestId);
+                               return CompletableFuture.completedFuture(null);
+                           }
+                           if (rgConfigRecord.getObject().isUpdating()) {
+                               if (isTransition) {
```

Comment:
i dont fully follow the logic here.. if it is a transition from non subscriber to subscriber then we do the addition.. so shouldnt we remove it when the transition is from subscriber to non subscriber too?

Reply:
The only time isTransition = "false" is when the Retention.Type="NONE" in both the old and new config. In this case there is no need to change anything in StreamMetadata (no subscriber adds/deletes)
In all other cases, isTransition = true and so we add/remove RGs to Stream Metadata as needed.

Reply:
thats the point -- if transition is from NONE to SUBSCRIPTION then the above logic works fine.. 
but when the transition is from subscription to NONE.. do you still end up adding the RG for subscription to streams that are present in the config or are all streams unsubscribed from this RG?

Reply:
The if check at ln # 89 guards against that. 
So it works both ways. I have added tests for all these transitions in StreamMetadataTasksTest.